### PR TITLE
rd/aws_instance and rd/aws_launch_template: Add support for metadata_options

### PR DESCRIPTION
**Pull request by stefansundin**
This adds support to set the Metadata Options options for an EC2 instance. TL;DR: opt-in to a more secure metadata service.

Documentation:
- https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html
- https://docs.aws.amazon.com/cli/latest/reference/ec2/run-instances.html
- https://docs.aws.amazon.com/cli/latest/reference/ec2/modify-instance-metadata-options.html

I am hopeful that they will add support for this to launch templates as well, but it seems like this has not happened yet. We can wait a few weeks to see if it comes out after reInvent.

Please let me know what you want to be changed. I'm not sure to what extent you want this to be tested. I've been testing this a bit and it is working well. I was unsure if you wanted the `state` attribute to be made available?

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/aws_instance: Add support for `metadata_options` parameter
data-source/aws_instance: Add `metadata_options` attributes
resource/aws_launch_template: Add support for `metadata_options` parameter
data-source/aws_launch_template: Add `metadata_options` attributes
```

Output from acceptance testing:
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSInstance_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSInstance_basic -timeout 120m
=== RUN   TestAccAWSInstance_basic
=== PAUSE TestAccAWSInstance_basic
=== CONT  TestAccAWSInstance_basic
--- PASS: TestAccAWSInstance_basic (162.39s)
PASS
ok    github.com/terraform-providers/terraform-provider-aws/aws 164.053s

$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSInstanceDataSource_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSInstanceDataSource_basic -timeout 120m
=== RUN   TestAccAWSInstanceDataSource_basic
=== PAUSE TestAccAWSInstanceDataSource_basic
=== CONT  TestAccAWSInstanceDataSource_basic
--- PASS: TestAccAWSInstanceDataSource_basic (253.76s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	255.474s
```

### CODE REVIEW

**Comment by stefansundin** [comment]
It was confirmed in the SEC310 reInvent session that launch template support is due in the next few weeks: https://youtu.be/2B5bhZzayjI?t=1966

**Comment by stefansundin** [comment]
It seems that the launch template support was launched without being announced anywhere (at least not as far as I could see). I just looked for it in the API docs today, and I saw that it had been added. I spent some time to add the launch template support, so I think this PR is ready for review now.

My `launch_template` test is failing with this error:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSLaunchTemplate_data'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSLaunchTemplate_data -timeout 120m
=== RUN   TestAccAWSLaunchTemplate_data
=== PAUSE TestAccAWSLaunchTemplate_data
=== CONT  TestAccAWSLaunchTemplate_data
--- FAIL: TestAccAWSLaunchTemplate_data (15.84s)
    testing.go:640: Step 0 error: Check failed: Check 13/20 error: aws_launch_template.test: Attribute 'metadata_options.#' expected "1", got "0"
FAIL
FAIL  github.com/terraform-providers/terraform-provider-aws/aws 17.508s
FAIL
make: *** [testacc] Error 1
```

I suspect this is because the metadata options are "pending" before fully applied to the launch template? It may be an error (i.e. copy-paste from the EC2 instances API), since there's something called `LaunchTemplateInstanceMetadataOptionsStatePending` in the SDK docs. It doesn't make a whole lot of sense to me since it shouldn't need to update a running thing, just publish a new version of the launch template.

Anyway, let me know what to do here. Thanks.

File: aws/resource_aws_instance.go [hunk 1]
```diff
@@ -1838,6 +1901,19 @@ func buildAwsInstanceOpts(
...
+		if mo, ok := v.([]interface{})[0].(map[string]interface{}); ok {
+			opts.MetadataOptions = &ec2.InstanceMetadataOptionsRequest{
+				HttpEndpoint: aws.String(mo["http_endpoint"].(string)),
+			}
+			if mo["http_endpoint"].(string) == ec2.InstanceMetadataEndpointStateEnabled {
+				// These parameters are not allowed unless HttpEndpoint is enabled
+				opts.MetadataOptions.HttpTokens = aws.String(mo["http_tokens"].(string))
+				opts.MetadataOptions.HttpPutResponseHopLimit = aws.Int64(int64(mo["http_put_response_hop_limit"].(int)))
+			}
+		}
```
Comment (bflad) [inline aws/resource_aws_instance.go]: Similar to creating a flatten function above, the inverse expand function could be created and used in the multiple places with this logic:

```go
func expandEc2InstanceMetadataOptions(l []interface{}) *ec2.MetadataOptions {
	m, ok := l[0].(map[string]interface{})

	if !ok {
		return nil
	}

	opts := &ec2.InstanceMetadataOptionsRequest{
		HttpEndpoint: aws.String(m["http_endpoint"].(string)),
	}

	if m["http_endpoint"].(string) == ec2.InstanceMetadataEndpointStateEnabled {
		// These parameters are not allowed unless HttpEndpoint is enabled

		if v, ok := m["http_tokens"].(string); ok && v != "" {
			opts.HttpTokens = aws.String(v)
		}

		if v, ok := m["http_put_response_hop_limit"].(int); ok && v != 0 {
			opts.HttpPutResponseHopLimit = aws.Int64(int64(v))
		}
	}

	return opts
}

// ...

MetadataOptions: expandEc2InstanceMetadataOptions(d.Get("metadata_options").([]interface{})),
```

File: aws/resource_aws_instance_test.go [hunk 2]
```diff
@@ -2814,6 +2817,12 @@ resource "aws_instance" "test" {
   instance_type   = "m1.small"
   security_groups = ["${aws_security_group.tf_test_test.name}"]
   user_data       = "foo:-with-character's"
+
+  metadata_options {
```
Comment (bflad) [inline aws/resource_aws_instance_test.go]: Please create a new test function and test configuration so we can verify existing testing for regressions.

File: aws/resource_aws_launch_template_test.go [hunk 3]
```diff
@@ -978,6 +979,12 @@ resource "aws_launch_template" "test" {
 
   key_name = "test"
 
+  metadata_options {
```
Comment (bflad) [inline aws/resource_aws_launch_template_test.go]: Same here about creating a separate test function and configuration.

**Comment by ewbankkit** [comment]
Closes https://github.com/terraform-providers/terraform-provider-aws/issues/10949.
Closes https://github.com/terraform-providers/terraform-provider-aws/issues/11794.

**Comment by ewbankkit** [comment]
@stefansundin Will you have a chance to address the review comments?
I have interest in using this functionality and can help out/complete if necessary.
Thanks.

**Comment by ewbankkit** [comment]
@stefansundin I have addressed @bflad's review comments via https://github.com/terraform-providers/terraform-provider-aws/pull/12491.

**Pull request closed without merging**